be/src/util/exception.c

```diff
@@ -0,0 +1,225 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Copied from:
+//   https://github.com/apache/hadoop/blob/release-2.7.1/hadoop-hdfs-project/hadoop-hdfs/src/main/native/libhdfs/exception.c
+// We need Hadoop 2.7 for Hortonworks HDP support, but libhdfs no longer exports
+// getJNIEnv so we have to copy some source files like this to Impala. ROCANA-3465
```

Comment:
And you're right, we do still link against libhdfs, but I don't think we'll run into symbol collisions.

In Hadoop 2.7 they made a change where they no longer publicly export methods by default. In fact, they still have the `getJNIEnv` function that we need, we just can't call it anymore.

Since we got an error when we tried to calling the now private `getJNIEnv` I think it means we can't have symbol collisions.

The way I initially attacked this was to pull in just the methods I needed, so as I got an error I would pull in more code, and as I could see that code calling other methods I would pull those methods in too. This got tedious so eventually I just started pulling in the whole files (which is what I did for thread_local_storage.c and htable.c), but even then, I still got some error that a method in the file wasn't public in libhdfs.

So, I think we're good, but it would be safe if i review the methods in this commit just to make sure they aren't exported.